src/Microsoft.AspNetCore.SignalR.Client/HubConnectionBuilderHttpExtensions.cs

```diff
@@ -62,7 +74,7 @@ public static IHubConnectionBuilder WithTransport(this IHubConnectionBuilder hub
             return hubConnectionBuilder;
         }
 
-        public static IHubConnectionBuilder WithMessageHandler(this IHubConnectionBuilder hubConnectionBuilder, HttpMessageHandler httpMessageHandler)
+        public static IHubConnectionBuilder WithMessageHandler(this IHubConnectionBuilder hubConnectionBuilder, Func<HttpMessageHandler, HttpMessageHandler> httpMessageHandler)
```

Comment:
Was chatting with @davidfowl and @Tratcher about this. It sounds like we have some precedent for the model where we give you the individual properties (`Cookies`, `Headers`, etc.) **and** a property to replace the `HttpMessageHandler` itself and if you set the latter, we **ignore** the individual properties (it's the "take my life into my own hands" option). I think we should stick with that for now.

The tricky thing here is that when we people do replace the `HttpMessageHandler`, they basically have to replace the `ClientWebSocketOptions` as well.

The error messages seem a little unpleasant for this too, since the user is providing a delegate and could return null in the delegate (so we can't just have an ArgumentNullException here).

Reply:
Whichever we choose, this could benefit from ye-olde-`///` comments describing the behavior a little bit.

Reply:
`HttpMessageHandler`s compose. A func is useful because devs don't necessarily want to completely replace the handler, they just want to wrap the inner `HttpMessageHandler` - which has been initialized with the other options on `HttpOptions` - with a handler that logs or something and then calls the inner handler.

I should also add that the func turned out to be useful for unit testing the new settings https://github.com/aspnet/SignalR/pull/1557/commits/3e8604670a4a3ea12ab12ddf4735ad78008a468b#diff-b5ac650137bfa6943ef9d5d0cfb05460R169

Reply:
Docs added

Reply:
Fair enough, let's go with it. We've got some time to get feedback anyway :)